### PR TITLE
Fix memcpy byte-to-char loop reduction for little endian architectures

### PR DESCRIPTION
**Pull request by 0xdaryl**
Idiom recognition is incorrectly transforming byte-to-char loops for little endian.  For example,
```
    for (int i = 0, j = 0; j < numChars; i += 2, j++) {
        dst[j + offsetDst] = (char) (((src[i + 1 + offsetSrc] & 0xFF) << 8) | (src[i + offsetSrc] & 0xFF))
    }
```
A loop such as this can be recognized and reduced to an arraycopy on little endian architectures.  Similarly, if the char was composed with the shift on the lower address byte then it would be recognized as an arraycopy on big endian architectures.

However, for little endian, the wrong address tree is chosen to start the arraycopy (i.e., `src[i + 1 + offsetSrc]` is chosen instead of  `src[i + offsetSrc]`).  Big endian selects the correct address.

Select the starting address based on the endianess of the target platform.

The loop reduction unit tests for this idiom only provide big endian patterns.  A second test is provided to have the arraycopy pattern recognized on little endian platforms.

### CODE REVIEW

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk8,jdk11

**Comment by 0xdaryl** [comment]
@sajinakandy fyi

**Comment by 0xdaryl** [comment]
Jenkins test sanity osx jdk8

Artifactory timeout on the previous build.

**Comment by 0xdaryl** [comment]
Jenkins test sanity aix jdk8

**Comment by 0xdaryl** [comment]
Jenkins test sanity osx jdk8

Previous failure is the familiar backwards arraycopy crash.

**Comment by 0xdaryl** [comment]
@jdmpapin @joransiu : I'm not sure who the idiom recognition experts are, but since you've both had experience in the code I'm wondering if you would mind reviewing this minor fix please?

**Comment by 0xdaryl** [comment]
Un-WIPing as testing looks good.

**Review by joransiu — APPROVED** [review]
Good catch!   LGTM besides question on the copyright date for the new test.

**Comment by jdmpapin** [comment]
The transformer has [existing logic to determine `inputNode` in the byte-to-char case](https://github.com/eclipse-openj9/openj9/blob/8c1a66bc22270cf3cbc794148fb0edc76ebe830b/runtime/compiler/optimizer/IdiomTransformations.cpp#L6164-L6174). The `checkByteToChar()` function, if it succeeds, will set `inputNode` (which is passed by reference) to a value of its choosing, and the new value depends only on the other arguments, not on the original value of `inputNode`. This does actually find the correct starting address

The problem seems to be that this logic is not running at all. It's trying to detect the byte-to-char case by looking for a 16-bit unsigned array store (i.e. `cstorei`), but stores are sign-agnostic, and we've been trying to eliminate the "unsigned" variants of sign-agnostic opcodes. I expect that for code similar to the new test we would have previously generated `cstorei`, but at some point we started to generate `sstorei` instead. Removing the unsigned condition allows `checkByteToChar()` to find the right starting point without any changes to the important nodes.
```diff
--- a/runtime/compiler/optimizer/IdiomTransformations.cpp
+++ b/runtime/compiler/optimizer/IdiomTransformations.cpp
@@ -6163,7 +6163,7 @@ CISCTransform2ArrayCopyB2CorC2B(TR_CISCTransformer *trans)

    // check whether the transformation is valid
    //
-   if (outputMemNode->getOpCode().isShort() && outputMemNode->getOpCode().isUnsigned())
+   if (outputMemNode->getOpCode().isShort())
       {
       TR::Node * iorNode = trans->getP2TRepInLoop(P->getImportantNode(2))->getHeadOfTrNodeInfo()->_node;
       if (!checkByteToChar(comp, iorNode, inputNode, bigEndian))
```
Such a removal should be very safe as well, because that condition is meaningless

Re-enabling `checkByteToChar()` also re-enables other safety checks which we have been skipping. I haven't inspected the code closely enough to determine whether these are important, but they might be

File: test/functional/JIT_Test/src/jit/test/loopReduction/littleEndianByte2CharMemCpy.java
Comment (jdmpapin) [inline test/functional/JIT_Test/src/jit/test/loopReduction/littleEndianByte2CharMemCpy.java:49]: I realize that this is largely a copy of the big-endian version of the test, but I have some questions nonetheless. I'm fine if the answer for the moment is that at least it's an attempt to run the code, and that's better than no such attempt

Do we do anything to ensure that this method is compiled at a high-enough opt level? I think that the pattern is only enabled at hot and above

Do we / should we check that the loop was transformed or not transformed as expected?

Do we / should we check the results in `array2`?

Comment (0xdaryl) [reply test/functional/JIT_Test/src/jit/test/loopReduction/littleEndianByte2CharMemCpy.java:49]: You are right to have questions about these tests and their usefulness.  I did as well, which is why I opened #13144 to hopefully improve on that.  The scope of that work may go beyond this particular transform (covering other loop reduction tests).  My goal with adding this test to this PR was to mirror what test was already there but for little endian.

**Comment by 0xdaryl** [comment]
My apologies--this PR fell off my radar.  I believe @jdmpapin's analysis reveals the proper fix in the idiom transformer to enable the logic that is already there to detect the correct starting address for little endian.  This was likely a bug introduced when the character type opcodes were retired.  Thanks for the careful review Devin.  I will push a new commit with that rather simple fix.

File: test/functional/JIT_Test/src/jit/test/loopReduction/littleEndianByte2CharMemCpy.java
Comment (0xdaryl) [reply test/functional/JIT_Test/src/jit/test/loopReduction/littleEndianByte2CharMemCpy.java]: Fixed.

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk8,jdk17

**Comment by 0xdaryl** [comment]
Jenkins test sanity aix jdk17

**Review by jdmpapin — APPROVED** [review]

**Comment by 0xdaryl** [comment]
Jenkins test sanity aix jdk8

**Comment by 0xdaryl** [comment]
Jenkins test sanity osx jdk17

**Comment by 0xdaryl** [comment]
Jenkins test sanity aix jdk8

Failure seems infra related.  One more try.

**Comment by 0xdaryl** [comment]
Jenkins test sanity aix jdk8

**Comment by 0xdaryl** [comment]
I have updated the PR to address all review comments.  PR testing has completed.  The linter job failure is expected.

@joransiu : would you mind being the committer on this one?  I will also open a 0.29 PR for this after this merges.

**Review by joransiu — APPROVED** [review]
Reviewed the latest changes. LGTM.

**Pull request merged**
